### ISSUE TITLE
how to check the wer summary result?

### ISSUE THREAD

**Issue by zcy618** [athena-team/athena#237]
dear friends:
I think I have successfully finished the asr/aishell example training and decoding, but I do not know where to check the summary result.
as you said:

Transformer (Ours without pre-training) | 6.82 | 57.88M
-- | -- | --
Transformer (Ours with self pre-training) | 6.6 | 57.88M

but from my log, I could only see every testing sentence like below:
INFO:absl:predictions: tf.Tensor(
[[2172 1379   38 1600 3159 1831  168 1946  801 2735  816 2079 1695  111
  1404 2554 2580 1600 3387 1429 4233]], shape=(1, 21), dtype=int64)	labels: [[2172 1379   38 1600 3159 1831  168 1946  801 2735  816 2079 1695  111
  2272 2554 2580 1600 3387 1429 2172 1379  980 2991 1786  705  306 2758
  1187 1799 1019   96 1018  145 3118 1946]]	errs: 17	avg_acc: 0.9210	sec/iter: 158.2585
INFO:absl:predictions: tf.Tensor([[ 237  273  299  168 2349 2724  357 3571 2184  575 4233]], shape=(1, 11), dtype=int64)	labels: [[ 237  273  299  168 2349 2724 3550 3571 2184  575 2498 2716 3755  143
  3571 2184  206  941  163 2856]]	errs: 11	avg_acc: 0.9209	sec/iter: 91.5847
INFO:absl:predictions: tf.Tensor(
[[2551   74  423  566 1755 2574   38    3 1772 1979 1348 1195 3595    3
   566 1755 2574   38    3 1772 1979 4233]], shape=(1, 22), dtype=int64)	labels: [[2551   74  423 3844 3249 2227 2787    3 1772 1979 2551   74  423 3844
  3249 2227 2787    3 1772 2554 1772 1979 3613 3449 2994 1440 1647]]	errs: 19	avg_acc: 0.9208	sec/iter: 188.1565
INFO:absl:decoding finished

but I do not know where where is the summary result.

And I notice the step 6 in run.sh:
if [ ${stage} -le 6 ] && [ ${stop_stage} -ge 6 ]; then
    # score-computing stage
    echo "computing score with sclite ..."
    if ! [ -x "$(command -v sclite)" ]; then
        echo "sclite is not installed !!"
        exit 1
    else
        . ./athena/tools/score_computer.sh decode.log examples/asr/aishell/data/vocab score_save || exit 1
    fi
fi

and I could see the folder: athena/score_save, but there is not what I want. So could you help to tell me how to understand your result please?
Thanks.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by Some-random** [comment]
The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?

Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

**Comment by zcy618** [comment]
> The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?
> 
> Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.

yes, I have install sclite, and run step 6, but there is only one log:
computing score with sclite ...

Then nothing else. so what do I miss?

**Comment by zcy618** [comment]
I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
Thanks

**Comment by Some-random** [comment]
> I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> Thanks

You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.

**Comment by Some-random** [comment]
> > The last avg_acc in decoding result file is the overall accuracy of test set because we accumulate errors and total characters during the decoding of test set. Your accuracy seems pretty low (92.08), did you run it using four gpu cards and the batch size we specify in example json?
> > Kaldi-style WER display is also supported in step 6 and it's more clear. To use it, you need to install sclite and run step 6 again.
> 
> yes, I have install sclite, and run step 6, but there is only one log:
> computing score with sclite ...
> 
> Then nothing else. so what do I miss?

We create a  folder called **score_save** in step 6 for score files. Please check files inside that folder

**Comment by zcy618** [comment]
![image](https://user-images.githubusercontent.com/29799298/90507692-e317a180-e188-11ea-955b-305b492c639b.png)

Yes, I noticed it, but I do not understand what I can get from it
[tmp1.txt](https://github.com/athena-team/athena/files/5090057/tmp1.txt)
[tmp2.txt](https://github.com/athena-team/athena/files/5090058/tmp2.txt)

I changed *.list to *.txt in order to upload them, could you help to check what should I check from them?
Thanks

**Comment by zcy618** [comment]
> > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > Thanks
> 
> You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.

if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
Thanks.

**Comment by Some-random** [comment]
> ![image](https://user-images.githubusercontent.com/29799298/90507692-e317a180-e188-11ea-955b-305b492c639b.png)
> 
> Yes, I noticed it, but I do not understand what I can get from it
> [tmp1.txt](https://github.com/athena-team/athena/files/5090057/tmp1.txt)
> [tmp2.txt](https://github.com/athena-team/athena/files/5090058/tmp2.txt)
> 
> I changed *.list to *.txt in order to upload them, could you help to check what should I check from them?
> Thanks

Can you also upload your decode.log please?

**Comment by Some-random** [comment]
> > > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > > Thanks
> > 
> > 
> > You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.
> 
> if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
> Thanks.

If you used **mtl_transformer_sp.json** with 8 gpu cards and a total batch size of (8 * 32), you should be able to at least get similar results as ours. Can you upload your training log please?

**Comment by zcy618** [comment]
> > ![image](https://user-images.githubusercontent.com/29799298/90507692-e317a180-e188-11ea-955b-305b492c639b.png)
> > Yes, I noticed it, but I do not understand what I can get from it
> > [tmp1.txt](https://github.com/athena-team/athena/files/5090057/tmp1.txt)
> > [tmp2.txt](https://github.com/athena-team/athena/files/5090058/tmp2.txt)
> > I changed *.list to *.txt in order to upload them, could you help to check what should I check from them?
> > Thanks
> 
> Can you also upload your decode.log please?

the decode.list is zero, nothing content, so I do not upload it.
Thanks

**Comment by zcy618** [comment]
> > > > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > > > Thanks
> > > 
> > > 
> > > You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.
> > 
> > 
> > if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
> > Thanks.
> 
> If you used **mtl_transformer_sp.json** with 8 gpu cards and a total batch size of (8 * 32), you should be able to at least get similar results as ours. Can you upload your training log please?


[nohup.out.zip](https://github.com/athena-team/athena/files/5090100/nohup.out.zip)

Yes, sure, it is here.

**Comment by Some-random** [comment]
> > > ![image](https://user-images.githubusercontent.com/29799298/90507692-e317a180-e188-11ea-955b-305b492c639b.png)
> > > Yes, I noticed it, but I do not understand what I can get from it
> > > [tmp1.txt](https://github.com/athena-team/athena/files/5090057/tmp1.txt)
> > > [tmp2.txt](https://github.com/athena-team/athena/files/5090058/tmp2.txt)
> > > I changed *.list to *.txt in order to upload them, could you help to check what should I check from them?
> > > Thanks
> > 
> > 
> > Can you also upload your decode.log please?
> 
> the decode.list is zero, nothing content, so I do not upload it.
> Thanks

It seems like you have written everything in nohup.out file but decode.log is supposed to be a separate file. You can take line 76236 to line 94744 from your nohup.out, make a decode.log yourself and run step 6 again.

**Comment by Some-random** [comment]
> > > > > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > > > > Thanks
> > > > 
> > > > 
> > > > You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.
> > > 
> > > 
> > > if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
> > > Thanks.
> > 
> > 
> > If you used **mtl_transformer_sp.json** with 8 gpu cards and a total batch size of (8 * 32), you should be able to at least get similar results as ours. Can you upload your training log please?
> 
> [nohup.out.zip](https://github.com/athena-team/athena/files/5090100/nohup.out.zip)
> 
> Yes, sure, it is here.

I've read the log, the batch size you used for ASR training is actually 16, but with 8 gpu cards it should be the same as our default config. Can you please check your cmvn file (examples/asr/aishell/data/cmvn)? Do you have per-speaker cmvn extracted for Aishell?

I also couldn't find your epoch 0 dev result for Aishell. Looking at the log, it seems the ASR training starts from epoch 1, which isn't right.

**Comment by zcy618** [comment]
> > > > > > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > > > > > Thanks
> > > > > 
> > > > > 
> > > > > You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.
> > > > 
> > > > 
> > > > if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
> > > > Thanks.
> > > 
> > > 
> > > If you used **mtl_transformer_sp.json** with 8 gpu cards and a total batch size of (8 * 32), you should be able to at least get similar results as ours. Can you upload your training log please?
> > 
> > 
> > [nohup.out.zip](https://github.com/athena-team/athena/files/5090100/nohup.out.zip)
> > Yes, sure, it is here.
> 
> I've read the log, the batch size you used for ASR training is actually 16, but with 8 gpu cards it should be the same as our default config. Can you please check your cmvn file (examples/asr/aishell/data/cmvn)? Do you have per-speaker cmvn extracted for Aishell?
> 
> I also couldn't find your epoch 0 dev result for Aishell. Looking at the log, it seems the ASR training starts from epoch 1, which isn't right.

Firstly, thanks very much for your log checking, I think maybe there are some breaks during the training because some unexpected errors. So do you think it is helpful if I git pull the latest code and re-run aishell: run.sh? If so, do you have some suggestions before I run it please, for examples, how to capture log, and how to run(earlier, I used nohup ./example/asr/aishell/run.sh &).
Thanks.

**Comment by Some-random** [comment]
> > > > > > > I used 8 gpu card to run, but because memory limitation, I set the batch size is 32.
> > > > > > > Thanks
> > > > > > 
> > > > > > 
> > > > > > You need a total batch size of 128 (batch size = 32 for 4 gpu cards or batch size = 16 for 8 gpu cards) to get optimal results because Transformer models get better results with larger batch size.
> > > > > 
> > > > > 
> > > > > if so, I am using 8gpu with batch size 32, I think it should get better result, actually it is not, what should I do next?
> > > > > Thanks.
> > > > 
> > > > 
> > > > If you used **mtl_transformer_sp.json** with 8 gpu cards and a total batch size of (8 * 32), you should be able to at least get similar results as ours. Can you upload your training log please?
> > > 
> > > 
> > > [nohup.out.zip](https://github.com/athena-team/athena/files/5090100/nohup.out.zip)
> > > Yes, sure, it is here.
> > 
> > 
> > I've read the log, the batch size you used for ASR training is actually 16, but with 8 gpu cards it should be the same as our default config. Can you please check your cmvn file (examples/asr/aishell/data/cmvn)? Do you have per-speaker cmvn extracted for Aishell?
> > I also couldn't find your epoch 0 dev result for Aishell. Looking at the log, it seems the ASR training starts from epoch 1, which isn't right.
> 
> Firstly, thanks very much for your log checking, I think maybe there are some breaks during the training because some unexpected errors. So do you think it is helpful if I git pull the latest code and re-run aishell: run.sh? If so, do you have some suggestions before I run it please, for examples, how to capture log, and how to run(earlier, I used nohup ./example/asr/aishell/run.sh &).
> Thanks.

I think you only need to rerun stage 3 and stage 5 if your cmvn is correct. Can you please upload your cmvn file for aishell?

**Comment by zcy618** [comment]
Sure, it is here:

[cmvn.txt](https://github.com/athena-team/athena/files/5100223/cmvn.txt)

**Comment by Some-random** [comment]
The cmvn file seems alright. You can just remove all ckpt files in checkpoint_dir and just rerun step 3 and step 5. That should lead you to our published result.
